libraries/standard.js

```diff
@@ -1,83 +1,110 @@
+      }
     })
-    return new_data;
+    new_data.push(obj);
+  })
+  return new_data;
 }
 
 
 // check if read values are null, undefined or blank (empty string)
 // dst: destination object, data: data item being processed, 
 // data_field: field being processed, skip_errors: boolean, errors: array of bad data items
-const null_check = (dst, data, data_field, skip_errors, errors) => {
+const add_required = (dst, dst_field, data, data_field, skip_errors, errors) => {
   // data is null, undefined or an empty object
   if (!data || Object.keys(data).length === 0) {
     throw new Error("No data or data field given")
   }
   if (data_field) {
-    dst = data_field
+    dst[dst_field] = data_field
     return true
   }
   // null data_field and skip errors
   if (!data_field && skip_errors) {
-    console.log(`A required value was undefined in the data. The field ${data_field} was skipped.`)
+    console.log(`A required field ${dst_field} was undefined or null. Skipping entry`)
   } else {
     // null data_field and no skip errors
-    errors.push(data)
+    let err = {missing_field: dst_field, data_entry: data}
+    errors.push(err);
   }
-  return false  
+  return false
 }
 
 
 // takes the whole details object and removes any null values
 // use case: item.details = remove_if_null(details);
-const remove_if_null = function(details_obj){
+const remove_if_null = function (details_obj) {
   // go through the details object and remove any null, undefined, length 0, and if its a string with only white spaces (trim begin and end spaces, and if length is 0 then we know that the string is just white spaces)
   Object.keys(details_obj).forEach(key => {
-      if (details_obj[key] === null || details_obj[key] === undefined || details_obj[key].length <= 0 || details_obj[key].trim().length === 0) {
-        delete details_obj[key];
-      }
-    });
-    // return the details object without null values
-    return details_obj
+    if (details_obj[key] === null || details_obj[key] === undefined || details_obj[key].length <= 0 || details_obj[key].trim().length === 0) {
+      delete details_obj[key];
+    }
+  });
+  // return the details object without null values
+  return details_obj
 }
 
-const add_if_not_null = (dst, src) => {
+const add_if_not_null = (dst, dst_field, src) => {
   if (src) {
-    dst = src
+    dst[dst_field] = src
     return true
   }
   return false
 }
 
+const validate_params = function (params) {
+  // handle no passed options
+  if (!params) {
+    let empty_params = {};
+    empty_params.stringify = false;
+    empty_params.skip_errors = false;
+    empty_params.validate = false;
+    return empty_params;
+  }
+  
+  // validation with schema
+  if (params.schema && !params.validator) {
+    throw "schema provided with no validator";
+  } else if (!params.schema && params.validator) {
+    throw "validator provided with no schema";
+  } else if (params.schema && params.validator) {
+    schema = JSON.parse(params.schema);
```

Comment:
should this be params.schema = JSON.parse(params.schema) ?

Reply:
yes it should be, great catch